### PR TITLE
feat(checkout): INT-951 Implement Masterpass button in cart and quick

### PR DESCRIPTION
**Pull request by cmedinilla**
## What? [INT-951](https://jira.bigcommerce.com/browse/INT-951)
Create checkout-buttons masterpass strategy
&
Hide the shipping address in the customer section as requested in [INT-984](https://jira.bigcommerce.com/browse/INT-984)
## Why?
We need this strategy to implement the cart and quick cart buttons for Masterpass

Due the delivery dates i made only one PR in order to simplify the process

## Testing / Proof
image
image
image
image

### CODE REVIEW

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+                this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(methodId)),
+                this._store.dispatch(this._checkoutActionCreator.loadDefaultCheckout()),
+            ]
+        );
+
+        return states.then(state => {
+                const statePaymentMethod = state[0];
+                const stateCheckout = state[1];
+
+                if (this._isInitialized) {
```
Comment (metalvegetarianoprogresivo) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: This should be the first thing to check in the initialize, so you don't do unnecessary calls if this is true.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+import {
+    InvalidArgumentError,
+    MissingDataError,
+    MissingDataErrorType
+} from '../../common/error/errors';
+import { PaymentMethod, PaymentMethodActionCreator } from '../../payment';
+import { Masterpass, MasterpassCheckoutOptions, MasterpassScriptLoader } from '../../payment/strategies/masterpass';
+import { CheckoutButtonInitializeOptions, CheckoutButtonOptions } from '../checkout-button-options';
+
+import CheckoutButtonStrategy from './checkout-button-strategy';
```
Comment (vmparra) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: Import from Index

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+import { MasterpassButtonInitializeOptions } from './masterpass-button-options';
```
Comment (vmparra) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: Merge with previous import

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
+import { Checkout, CheckoutActionCreator, CheckoutStore } from '../../checkout';
+import {
+    InvalidArgumentError,
+    MissingDataError,
+    MissingDataErrorType
+} from '../../common/error/errors';
+import { PaymentMethod, PaymentMethodActionCreator } from '../../payment';
+import { Masterpass, MasterpassCheckoutOptions, MasterpassScriptLoader } from '../../payment/strategies/masterpass';
+import { CheckoutButtonInitializeOptions, CheckoutButtonOptions } from '../checkout-button-options';
```
Comment (vmparra) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: Import from Index

**Comment by cmedinilla** [comment]
@vmparra @metalvegetarianoprogresivo ♻️ INT-910 merged this is rebased a ready for another review

**Comment by cmedinilla** [comment]
Closed #437 and changes added to this PR

**Comment by cmedinilla** [comment]
♻️ @vmparra  All comments was addressed

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
+import { CheckoutButtonInitializeOptions, CheckoutButtonOptions } from '../';
+import { Checkout, CheckoutActionCreator, CheckoutStore } from '../../checkout';
+import {
```
Comment (vmparra) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: 🍹 In this case, you can put them in one single line.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    InvalidArgumentError,
+    MissingDataError,
+    MissingDataErrorType
+} from '../../common/error/errors';
+import { bindDecorator as bind } from '../../common/utility';
+import { PaymentMethod, PaymentMethodActionCreator } from '../../payment';
+import {
```
Comment (vmparra) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: 🍹 Here too.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts (hunk 5 above)
Comment (cmedinilla) [reply src/checkout-buttons/strategies/masterpass-button-strategy.ts]: image
i thin is too long here in one line

Comment (cmedinilla) [reply src/checkout-buttons/strategies/masterpass-button-strategy.ts]: i think the line is too long for that:
image

**Comment by Sure-Round** [comment]
@davidchin @icatalina This PR is ready for review ♻️ Of the Integration team PR's this is the top priority one right now.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            return super.initialize(options);
+        }
+
+        return Promise.all(
+            [
+                this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(methodId)),
+                this._store.dispatch(this._checkoutActionCreator.loadDefaultCheckout()),
+            ]
+        ).then(([statePayment, stateCheckout]) => {
+                this._paymentMethod = statePayment.paymentMethods.getPaymentMethod(methodId);
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: You shouldn't have to retain `statePayment` and `stateCheckout`. Both are snapshots of the same state, but captured at different times; so because the checkout action is dispatched later, `stateCheckout` is fresher than `statePayment`. Anyway, my recommendation is to retrieve it directly from `store` so you know for sure you're dealing with the latest data.

```ts
const state = this._store.getState();

this._paymentMethod = state.paymentMethods.getPaymentMethod(methodId);
this._checkout = state.checkout.getCheckout();
```

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+            this._signInButton.parentNode.removeChild(this._signInButton);
+            this._signInButton = undefined;
+        }
+
+        return super.deinitialize(options);
+    }
+
+    private _createSignInButton(masterpassOptions: MasterpassButtonInitializeOptions): void {
+        const { container } = masterpassOptions;
+        const buttoncontainer = document.querySelector(`#${container}`);
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: Could you please use camel case for the two words? i.e.: `buttonContainer`.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        };
+    }
+
+    @bind
+    private _handleWalletButtonClick(): void  {
+        const payload = this._createMasterpassPayload();
+        this.masterpassClient.checkout(payload);
+    }
+
+    private get masterpassClient(): Masterpass {
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: Firstly, we actually don't allow computed properties in this project. `get/set` are forbidden because it is very easy to put logic in there that's computationally expensive or induces side effects (for more information, please see this [article](https://github.com/airbnb/javascript#accessors--no-getters-setters)).

Secondly, all private properties and methods need to be prefixed with `_`. This is because, although we're using TypeScript, the source ultimately gets compiled into JS and consumed by JS clients. So we need a way to indicate to the consumers that these properties and methods are not supposed to be used.

If you really want to go with this pattern, you need to name this method as `_getMasterClient`. But often we just inline the guard statement unless there's a real need (i.e.: you need to access the property multiple times).

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        if (!this._masterpassClient) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        return this._masterpassClient;
+    }
+
+    private get paymentMethod(): PaymentMethod {
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: Please see my previous comment regarding the use of `get` and the name of the method. If you do want to have a private method for retrieving the `PaymentMethod` object, you should retrieve the data directly from `store` instead. This is because keeping a separate reference is harder to maintain as you always have to remember to update the reference when the object gets stale.

```ts
private _getPaymentMethod(): PaymentMethod {
    const paymentMethod = state.paymentMethods.getPaymentMethod(this._methodId);

    if (!paymentMethod) {
        throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
    }

    return paymentMethod;
}
```

That'd also simply your `initialize` method.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+    @bind
+    private _handleWalletButtonClick(): void  {
+        const payload = this._createMasterpassPayload();
+        this.masterpassClient.checkout(payload);
+    }
+
+    private get masterpassClient(): Masterpass {
+        if (!this._masterpassClient) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: Can you please throw `NotInitializeError`? It is a more appropriate error to throw in this case.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+    private get paymentMethod(): PaymentMethod {
+        if (!this._paymentMethod) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        return this._paymentMethod;
+    }
+
+    private get checkout(): Checkout {
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: Please see my previous comment regarding `paymentMethod`.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+        if (!this._checkout) {
+            throw new MissingDataError(MissingDataErrorType.MissingCart);
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: You're missing `Checkout` object, not `Cart`.

File: src/payment/strategies/masterpass/masterpass.ts [hunk 14]
```diff
@@ -46,7 +46,6 @@ export interface MasterpassCheckoutOptions {
      * This optional parameter can be used to override the callbackUrl specified in the Merchant Portal.
      */
     callbackUrl?: string;
-
```
Comment (davidchin) [inline src/payment/strategies/masterpass/masterpass.ts]: Unnecessary diff. Could you please revert?

**Comment by cmedinilla** [comment]
@davidchin All comments was addressed only the store state part with the proposed approach is not working
i think is because the checkout doesn't exist where this strategy is called (cart and quick cart section)

My proposal is load only the state checkout to get the latest data and from there load the paymentMethod too

````javascript
        .then(stateCheckout => {
                this._stateCheckout = stateCheckout;

                return this._masterpassScriptLoader.load(this._getPaymentMethod().config.testMode)
                    .then(masterpass => {
                        this._masterpassClient = masterpass;
                        this._createSignInButton(masterpassOptions);
                    });
        })
        .then(() => super.initialize(options));
````

````javascript
    private _getPaymentMethod(): PaymentMethod {
        if (!this._methodId) {
            throw new InvalidArgumentError();
        }

        if (!this._stateCheckout) {
            throw new MissingDataError(MissingDataErrorType.MissingCheckout);
        }

        const paymentMethod = this._stateCheckout.paymentMethods.getPaymentMethod(this._methodId);
        if (!paymentMethod || !paymentMethod.initializationData.checkoutId) {
            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
        }

        return paymentMethod;
    }
````

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+        if (!masterpassOptions || !methodId) {
+            throw new InvalidArgumentError('Unable to proceed because "options.masterpass" argument is not provided.');
+        }
+        this._methodId = methodId;
+
+        if (this._isInitialized) {
+            return super.initialize(options);
+        }
+
+        return this._store.dispatch(this._checkoutActionCreator.loadDefaultCheckout())
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: Hey @cmedinilla, sorry if I didn't explain myself clearly last time, I didn't mean to remove `loadPaymentMethod` and `loadDefaultCheckout`. You still need those. All I'm asking is not to read it from `stateCheckout` and store a separate reference of it. So with what you've done so far, the code could be something like the following:

```ts
initialize(options: CheckoutButtonInitializeOptions): Promise <void> {
    const { masterpass: masterpassOptions, methodId } = options;

    if (!masterpassOptions || !methodId) {
        throw new InvalidArgumentError('Unable to proceed because "options.masterpass" argument is not provided.');
    }

    if (this._isInitialized) {
        return super.initialize(options);
    }

    this._methodId = methodId;

    return Promise.all([
        this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(methodId)),
        this._store.dispatch(this._checkoutActionCreator.loadDefaultCheckout()),
    ])
        .then(() => this._masterpassScriptLoader.load(this._getPaymentMethod().config.testMode))
        .then(masterpass => {
            this._masterpassClient = masterpass;
            
            this._createSignInButton(masterpassOptions);

            return super.initialize(options);
        });
}
```

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+
+        return this._masterpassClient;
+    }
+
+    private _getPaymentMethod(): PaymentMethod {
+        if (!this._methodId) {
+            throw new InvalidArgumentError();
+        }
+
+        if (!this._stateCheckout) {
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: This check should be removed, as the checkout object shouldn't affect the retrieval of the payment method object.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+        const paymentMethod = this._stateCheckout.paymentMethods.getPaymentMethod(this._methodId);
+        if (!paymentMethod || !paymentMethod.initializationData.checkoutId) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        return paymentMethod;
+    }
+
+    private _getCheckout(): Checkout {
+        if (!this._stateCheckout) {
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: This check should be removed as well. The check below is sufficient.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+        if (!this._masterpassClient) {
+            throw new NotInitializedError(NotInitializedErrorType.CheckoutButtonNotInitialized);
+        }
+
+        return this._masterpassClient;
+    }
+
+    private _getPaymentMethod(): PaymentMethod {
+        if (!this._methodId) {
+            throw new InvalidArgumentError();
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: In this case, you should throw `NotInitializedError` because first of all this method doesn't accept an argument so it's odd to throw `InvalidArgumentError`, secondly the reason why `methodId` could be undefined is because the strategy is not initialized.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+    MasterpassCheckoutOptions,
+    MasterpassScriptLoader
+} from '../../payment/strategies/masterpass';
+
+import { CheckoutButtonStrategy, MasterpassButtonInitializeOptions } from './';
+
+export default class MasterpassButtonStrategy extends CheckoutButtonStrategy {
+    private _masterpassClient?: Masterpass;
+    private _signInButton?: HTMLElement;
+    private _stateCheckout?: InternalCheckoutSelectors;
```
Comment (davidchin) [inline src/checkout-buttons/strategies/masterpass-button-strategy.ts]: Please see my comment above regarding this. It should be removed.

File: src/checkout-buttons/strategies/masterpass-button-strategy.ts (hunk 15 above)
Comment (cmedinilla) [reply src/checkout-buttons/strategies/masterpass-button-strategy.ts]: I made some changes, now i am not saving the state anymore and i am only using the data to create the payload

**Comment by cmedinilla** [comment]
@davidchin ♻️

**Comment by Sure-Round** [comment]
@cmedinilla If you squash commits I will merge for you.
